libclamav/pdf.c

```diff
@@ -1429,7 +1429,7 @@ static int pdf_scan_contents(int fd, struct pdf_struct *pdf)
     return rc;
 }
 
-cl_error_t pdf_extract_obj(struct pdf_struct *pdf, struct pdf_obj *obj, uint32_t flags)
+cl_error_t pdf_extract_obj(struct pdf_struct *pdf, struct pdf_obj *obj, uint32_t flags, uint32_t rec_level)
```

Comment:
You could avoid this API change by just making this function a wrapper, and then call the new (static) function with the initial recursion depth.
``` c
cl_error_t pdf_extract_obj(struct pdf_struct *pdf, struct pdf_obj *obj, uint32_t flags)
{
    return _pdf_extract_obj(pdf, obj, PRIVATE_STUFF);
}
```

You could further future-proof it by passing around a parsing context, which might be as small as this for now (but might track other resource-exhausting elements later):
``` c
typedef struct {
    int recursion_depth;
} PdfParseCxt;
```

It's small enough to pass by value. No need to allocate and pass by pointer.
``` c
cl_error_t pdf_extract_obj(struct pdf_struct *pdf, struct pdf_obj *obj, uint32_t flags, PdfParseCxt pcxt)
```

Also, I tend to think of (and see) this parameter as "depth" rather than "level".

Reply:
It's not really an API, but your point is a good one for any function that is very commonly used or for public API's.  
In this case I feel we're okay - `pdf_extract_obj()` and the `pdf_parse_*` functions in question aren't used very much and are purely internal to the pdf parser.  It's just in `pdf.h` because whoever extended the PDF parser didn't bother modularizing nicely with a `pdf_internal.h` file for things shared between the different C files. 👀 Pretty common design issue with a lot of the clamav parsers, honestly. Could definitely stand to be made into more tidy modules. 
Anyways, I see your point but I think we're okay on this one.